File: source/scala/src/Parallel.scala
// Parallel map with atomic work stealing

import java.util.concurrent.atomic.AtomicInteger

object Parallel:
  def map[T, R: reflect.ClassTag](
      cores: Int,
      init: R,
      run: T => R,
      tasks: Array[T]
  ): Array[R] =
    val count = tasks.length
    val next = AtomicInteger(0)
    val results = Array.fill(count)(init)

    // Shuffle work order for load balancing
    val order = Array.tabulate(count)(identity)
    val rng = java.util.concurrent.ThreadLocalRandom.current()
    var i = count - 1
    while i > 0 do
      val j = rng.nextInt(i + 1)
      val tmp = order(i); order(i) = order(j); order(j) = tmp
      i -= 1

    def worker(): Unit =
      var index = next.getAndIncrement()
      while index < count do
        val idx = order(index)
        results(idx) = run(tasks(idx))
        index = next.getAndIncrement()

    val threads = Array.tabulate(cores - 1) { _ =>
      val t = Thread((() => worker()): Runnable)
      t.start()
      t
    }

    worker()
    threads.foreach(_.join())
    results

File: source/kotlin/src/Parallel.kt
// Parallel map with atomic work stealing

import java.util.concurrent.ThreadLocalRandom
import java.util.concurrent.atomic.AtomicInteger

object Parallel {
    fun <T, R> map(cores: Int, tasks: Array<T>, run: (T) -> R): List<R> {
        val count = tasks.size
        val next = AtomicInteger(0)
        val results = arrayOfNulls<Any>(count)

        // Shuffle work order for load balancing
        val order = IntArray(count) { it }
        val rng = ThreadLocalRandom.current()
        for (i in count - 1 downTo 1) {
            val j = rng.nextInt(i + 1)
            val tmp = order[i]; order[i] = order[j]; order[j] = tmp
        }

        val worker = Runnable {
            var index = next.getAndIncrement()
            while (index < count) {
                val idx = order[index]
                results[idx] = run(tasks[idx])
                index = next.getAndIncrement()
            }
        }

        val threads = List(cores - 1) { Thread(worker).also { it.start() } }
        worker.run()
        threads.forEach { it.join() }
        @Suppress("UNCHECKED_CAST")
        return results.map { it as R }
    }
}
